include/rocksdb/table.h

```diff
@@ -104,6 +106,20 @@ struct MetadataCacheOptions {
   PinningTier unpartitioned_pinning = PinningTier::kFallback;
 };
 
+struct CacheEntryRoleOptions {
+  enum class Decision {
+    kEnabled,
+    kDisabled,
+    kFallback,
+  };
+  Decision charged = Decision::kFallback;
+};
+
+struct CacheUsageOptions {
+  CacheEntryRoleOptions default_options;
+  std::array<CacheEntryRoleOptions, kNumCacheEntryRoles> options_overrides;
```

Comment:
[NOTE]
I often find reasoning about whether data member's definition like this will be properly default-initialized (i.e, `{CacheEntryRoleOptions(), CacheEntryRoleOptions(), ..., CacheEntryRoleOptions()}`)

My reasoning for this one is that the trivial default constructor `CacheUsageOptions()` will always be called (cuz there is no other constructor) and according to the "situation 3)" in https://en.cppreference.com/w/cpp/language/default_initialization saying "when ... or a non-static data member is not mentioned in a [constructor initializer list](https://en.cppreference.com/w/cpp/language/constructor) and that constructor is called", I believe `options_overrides` will be such non-static data member and get default initialization. Then according to "if T is an array type, every element of the array is default-initialized;", I believe for each element, the trivial default constructor `CacheEntryRoleOptions()` will be called so we should be safe.

Does reasoning about things like this ever bother you? Is our CI strong enough catch undefined undetermined initialized value like this when it isn't properly default initialized?

Reply:
Sorry I meant to suggest std::array for the internal representation. Is there a way to still use `std::map` publicly and use an array-like representation internally? Otherwise we have to worry about preserving the value of existing enum values so introducing new CacheEntryRoles does not break existing configs. But that is difficult - one reason why is because kMisc keeps getting its value bumped as the doc says to add new enum values before it.

> Does reasoning about things like this ever bother you?

Yes, always. Since it has the intuitive behavior here it is OK with me for this point.

> Is our CI strong enough catch undefined undetermined initialized value like this when it isn't properly default initialized?

Don't think so. Last time I thought about it was here - https://github.com/facebook/rocksdb/pull/5720#pullrequestreview-276899350.

Reply:
> Sorry I meant to suggest std::array for the internal representation. Is there a way to still use std::map publicly and use an array-like representation internally? Otherwise we have to worry about preserving the value of existing enum values so introducing new CacheEntryRoles does not break existing configs.

Aha - yeah let me fix that quickly and figure out where to store this internal rep.